### PR TITLE
[sonic-cfggen] optimize sonic-cfggen startup

### PR DESCRIPTION
**Pull request by stepanblyschak**
**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by lguohan** [comment]
do you have a benchmark number?

**Comment by stepanblyschak** [comment]
@lguohan
[cfggen_prof.txt](https://github.com/Azure/sonic-buildimage/files/3766584/cfggen_prof.txt)
Test done with lazy_re. Note, sonic-cfggen is called over 100 times during start, this saves 8 sec for LAG restoration time. Ideally I think we could rethink the way start scripts work to eliminate many sonic-cfggen invocations.
Sonic-cfggen functionality could be implemented as a library that start scripts import only once (import jinja2, natsort are expensive because many static regular expressions are compiled at import time). This I believe will give a boost for start script like in bgp docker container https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-fpm-frr/start.sh where it renders many templates.
For scripts that require only hwsku, platform, hostname there could be a module that does not import heavy packages at all, since all you need is single redis query or read from machine.conf.
Some templates are questionable if they even have to be templates at all, like https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-dhcp-relay/wait_for_intf.sh.j2. A script could do this at runtime with few redis queries.
To speedup a bit templates rendering we can to use jinja cache in redis.
What do you think?

File: src/sonic-config-engine/sonic-cfggen [hunk 1]
```diff
@@ -16,6 +16,16 @@ See usage string for detail description for arguments.
 """
 
 from __future__ import print_function
+
+# mokey patch re.compile to do lazy regular expression compilation.
```
Comment (qiluo-msft) [inline src/sonic-config-engine/sonic-cfggen]: >mokey [](start = 2, length = 5)

typo

File: src/sonic-config-engine/lazy_re.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+        def __getattr__(self, name):
+            if self.pattern_obj is None:
+                self.pattern_obj = _orig_re_compile(*self.args, **self.kwargs)
+            return getattr(self.pattern_obj, name)
+    return  __LazyReCompile(*args, **kwargs)
+
+re.compile = __re_compile
+
+
```
Comment (qiluo-msft) [inline src/sonic-config-engine/lazy_re.py]: Several  code format issues. Please check [http://pep8online.com](http://pep8online.com)

File: src/sonic-config-engine/redis_bcc.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+class RedisBytecodeCache(jinja2.BytecodeCache):
+    """ A bytecode cache for jinja2 template that stores bytecode in Redis """
+
+    REDIS_HASH = 'JINJA2_CACHE'
+
+    def __init__(self, client):
+        self._client = client
+        try:
+            self._client.connect(self._client.STATE_DB, retry_on=False)
```
Comment (qiluo-msft) [inline src/sonic-config-engine/redis_bcc.py]: >STATE_DB [](start = 46, length = 8)

use another non-used database?

**Comment by qiluo-msft** [comment]
do you have a benchmark for RedisBytecodeCache improvement?

**Comment by lguohan** [comment]
@stepanblyschak , i agree with all comments. good ideas!

**Comment by lguohan** [comment]
please address other reviewers comments.

File: src/sonic-config-engine/redis_bcc.py (hunk 3 above)
Comment (stepanblyschak) [reply src/sonic-config-engine/redis_bcc.py]: Can I use db index 8 and update swss-common/swss-sdk?

**Comment by stepanblyschak** [comment]
[profiling_redis_bcc.txt](https://github.com/Azure/sonic-buildimage/files/3779594/profiling_redis_bcc.txt)

File: src/sonic-config-engine/redis_bcc.py (hunk 3 above)
Comment (qiluo-msft) [reply src/sonic-config-engine/redis_bcc.py]: This is still applicable.

---
In reply to: [338815799](https://github.com/Azure/sonic-buildimage/pull/3658#discussion_r338815799) [](ancestors = 338815799)

**Pull request merged**